File: templates/python/app/common/base_repository.py
import uuid
from typing import Generic, TypeVar

from sqlalchemy import desc, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.common.base_model import BaseModel
from app.common.pagination import PaginatedResult, PaginationMeta

T = TypeVar("T", bound=BaseModel)


class BaseRepository(Generic[T]):
    """Repositório base genérico."""

    def __init__(self, model: type[T], session: AsyncSession):
        self.model = model
        self.session = session

    async def find_all(self) -> list[T]:
        stmt = select(self.model).order_by(desc(self.model.created_at))
        result = await self.session.execute(stmt)
        return list(result.scalars().all())

    async def find_by_id(self, id: uuid.UUID) -> T | None:
        return await self.session.get(self.model, id)

    async def create(self, data: dict) -> T:
        entity = self.model(**data)
        self.session.add(entity)
        await self.session.flush()
        await self.session.refresh(entity)
        return entity

    async def update(self, id: uuid.UUID, data: dict) -> T | None:
        entity = await self.find_by_id(id)
        if not entity:
            return None
        for key, value in data.items():
            if value is not None:
                setattr(entity, key, value)
        await self.session.flush()
        await self.session.refresh(entity)
        return entity

    async def delete(self, id: uuid.UUID) -> None:
        entity = await self.find_by_id(id)
        if entity:
            await self.session.delete(entity)
            await self.session.flush()

    async def find_all_paginated(
        self,
        page: int = 1,
        limit: int = 10,
        order_by: str = "created_at",
        order_dir: str = "DESC",
    ) -> PaginatedResult[T]:
        # Contagem total
        count_stmt = select(func.count()).select_from(self.model)
        total_result = await self.session.execute(count_stmt)
        total_items = total_result.scalar() or 0

        # Query paginada
        column = getattr(self.model, order_by, self.model.created_at)
        order = desc(column) if order_dir.upper() == "DESC" else column.asc()

        stmt = select(self.model).order_by(order).offset((page - 1) * limit).limit(limit)
        result = await self.session.execute(stmt)
        data = list(result.scalars().all())

        total_pages = (total_items + limit - 1) // limit if limit > 0 else 0

        return PaginatedResult(
            data=data,
            meta=PaginationMeta(
                total_items=total_items,
                item_count=len(data),
                items_per_page=limit,
                total_pages=total_pages,
                current_page=page,
            ),
        )

File: templates/nest/src/common/base.repository.ts
import {
  DeepPartial,
  FindOptionsOrder,
  Repository,
  FindManyOptions,
  FindOptionsWhere,
} from 'typeorm';
import { QueryDeepPartialEntity } from 'typeorm/query-builder/QueryPartialEntity';
import { BaseEntity } from './base.entity';
import { IRepository } from './irepository';
import { buildPaginationMeta, PaginatedResult } from './utils/pagination.utils';

export abstract class BaseRepository<
  T extends BaseEntity,
> implements IRepository<T> {
  constructor(protected readonly repository: Repository<T>) {}

  public get repo(): Repository<T> {
    return this.repository;
  }

  async findAll(): Promise<T[]> {
    return this.repository.find({
      order: { createdAt: 'DESC' } as FindOptionsOrder<T>,
    });
  }

  async findById(id: string): Promise<T | null> {
    if (!id) {
      return null;
    }
    return this.repository.findOneBy({ id } as FindOptionsWhere<T>);
  }

  async save(entity: T): Promise<T> {
    return this.repository.save(entity);
  }

  async create(data: DeepPartial<T>): Promise<T> {
    const entity = this.repository.create(data);
    return this.repository.save(entity);
  }

  createInstance(data: DeepPartial<T>): T {
    return this.repository.create(data);
  }

  async update(id: string, data: QueryDeepPartialEntity<T>): Promise<T | null> {
    if (!id) {
      return null;
    }
    await this.repository.update(id, data);
    return this.findById(id);
  }

  async delete(id: string): Promise<void> {
    if (!id) {
      return;
    }
    await this.repository.delete(id);
  }

  async findAllPaginated(
    options: FindManyOptions<T>,
    meta: { page: number; limit: number },
  ): Promise<PaginatedResult<T>> {
    const { page, limit } = meta;
    const [data, totalItems] = await this.repository.findAndCount({
      ...options,
      skip: (page - 1) * limit,
      take: limit,
    });

    return {
      data,
      meta: buildPaginationMeta(totalItems, page, limit, data.length),
    };
  }
}
